torch_xla/distributed/parallel_loader.py

```diff
@@ -155,27 +157,64 @@ def _loader_worker(self):
       batch.append(data)
       if len(batch) == len(self._devices):
         for queue_no, device_batch in enumerate(batch):
-          queues[queue_no].loader_queue.put(device_batch)
+          queues[queue_no].cpu_loader_queue.put(device_batch)
         batch = []
     for dqueue in queues:
-      dqueue.loader_queue.close_write()
+      dqueue.cpu_loader_queue.close_write()
 
   def _get_batch(self, dqueue):
     batch = []
-    while dqueue.queue.max_size() > len(batch):
-      item = dqueue.loader_queue.get()
+    while len(batch) < dqueue.queue.max_size():
+      item = dqueue.cpu_loader_queue.get()
       if item is None:
         break
       batch.append(item)
     return batch
 
+  def send_cpu_data_to_device(self, batches, device):
+    """Move batch to device.
+
+    Args:
+      batch -> List(torch.Tensor), List(Dict(str: torch.Tensor)): Input batch
+        present in the cpu memory
+      device: TPU device where the batch should be moved
+    
+    Returns:
+      result -> List(torch.Tensor), Dict(str: torch.Tensor): Returns a dict if the
+        input batch is a dict. Otherwise, returns a list of torch.Tensor.
+    """
+    result = None
+    if isinstance(batches[0], dict) and isinstance(self._input_sharding, dict):
+      result = []
+      for batch in batches:
+        xla_batch = {}
+        missing_keys = []
+        for key, tensor in batch.items():
+          assert type(tensor) == torch.Tensor
+          if key not in self._input_sharding:
+            missing_keys.append(key)
+            continue
+          # xla_tensor is a list of tensors.
+          xla_tensor = xm.send_cpu_data_to_device(tensor, device,
+                                                  self._input_sharding[key])
+          xla_batch[key] = xla_tensor[0]
+        if len(missing_keys) != 0:
+          # Returning exception as raisng in the dataloading thread doesn't surface the problem in the main thread.
+          return [
+              KeyError(f"Keys: {missing_keys} are missing from input_sharding.")
+          ]
```

Comment:
why not raise directly?

Reply:
This part of the code executes in a separate thread. If an exception happens in the thread, the main thread is just blocked on the _worker thread to load data into the input_batches queue. To explicitly raise the error to the user, I think it might be better if we add the exception to the input batch queue and check for it while passing the data to the main thread in PerDeviceLoader.next function